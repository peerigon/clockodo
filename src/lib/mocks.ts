// These date helpers are not meant to replace an actual date library.
// They are inaccurate and sometimes just wrong, see https://gist.github.com/timvisee/fcda9bbdff88d45cc9061606b4b923ca
// They should only be used for generating mock data.

import { faker } from "@faker-js/faker";

/**
 * Sets the seed for Faker.js (which is used by the mocks).
 *
 * @see https://github.com/Marak/Faker.js
 */
export const setFakerSeed = (seed: number) => {
  faker.seed(seed);
};

/** The number of milliseconds on a typical day. */
export const ONE_DAY = 24 * 60 * 60 * 1000;

/** The number of milliseconds on a typical year. */
export const ONE_YEAR = 356 * ONE_DAY;

const MAX_ITERATION_COUNT = 10_000;

export const startOfDay = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
};

export const endOfMonth = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth() + 1, 0, 23, 59, 59, 999);
};

export const endOfYear = (dateTime: Date) => {
  return new Date(dateTime.getFullYear() + 1, 0, 0, 23, 59, 59, 999);
};

export const startOfNextDay = (date: Date) => {
  const nextDay = new Date(startOfDay(date).getTime() + 1.5 * ONE_DAY);

  return new Date(nextDay.getFullYear(), nextDay.getMonth(), nextDay.getDate());
};

const generateWithMaxDuplicates = <Value>({
  count,
  maxDuplicates = 1,
  generate,
}: {
  count: number;
  maxDuplicates?: number;
  generate: () => Value;
}) => {
  const accepted = new Map<Value, Array<Value>>();
  let acceptedCount = 0;

  for (let i = 0; acceptedCount < count; i++) {
    if (i === MAX_ITERATION_COUNT) {
      throw new Error(
        `Couldn't generate enough unique values before reaching the max iteration count.
This usually happens when the generated values create too many conflicts (e.g. too many dates on a small date range).`,
      );
    }
    const value = generate();
    const duplicates = accepted.get(value) ?? [];

    if (duplicates.length >= maxDuplicates) continue;

    duplicates.push(value);
    accepted.set(value, duplicates);
    acceptedCount++;
  }

  return [...accepted.values()].flat();
};

export const generateRandomDateTimes = ({
  count,
  between: [from, to],
  maxDuplicates = 1,
}: {
  count: number;
  between: readonly [Date, Date];
  maxDuplicates?: number;
}) => {
  return generateWithMaxDuplicates({
    count,
    maxDuplicates,
    generate: () => {
      return faker.date.between({ from, to }).getTime();
    },
  }).sort();
};

export const generateRandomDates = ({
  count,
  between: [from, to],
  maxDuplicates = 1,
}: {
  count: number;
  between: readonly [Date, Date];
  maxDuplicates?: number;
}) => {
  return generateWithMaxDuplicates({
    count,
    maxDuplicates,
    generate: () => {
      const randomDate = faker.date.between({ from, to });

      return new Date(
        randomDate.getFullYear(),
        randomDate.getMonth(),
        randomDate.getDate(),
      ).getTime();
    },
  }).sort();
};

export const generateRandomMonths = ({
  count,
  between: [from, to],
  maxDuplicates = 1,
}: {
  count: number;
  between: readonly [Date, Date];
  maxDuplicates?: number;
}) => {
  return generateWithMaxDuplicates({
    count,
    maxDuplicates,
    generate: () => {
      const randomDate = faker.date.between({ from, to });

      return new Date(
        randomDate.getFullYear(),
        randomDate.getMonth(),
      ).getTime();
    },
  }).sort();
};

export const toPairs = <Item>(array: Array<Item>) => {
  if (array.length % 2 !== 0) {
    throw new Error("Cannot create pairs: Array length must be divisible by 2");
  }

  const count = array.length / 2;
  const pairs: Array<[Item, Item]> = Array.from({ length: count });

  for (let i = 0; i < count; i++) {
    pairs[i] = [array[i * 2], array[i * 2 + 1]];
  }

  return pairs;
};
