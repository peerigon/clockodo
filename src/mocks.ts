import { faker } from "@faker-js/faker";

export * from "./models/absence.mocks.js";
export * from "./models/customer.mocks.js";
export * from "./models/entry.mocks.js";
// export * from "./models/entryGroup.mocks.js";
export * from "./models/holidayscarry.mocks.js";
export * from "./models/holidaysquota.mocks.js";
export * from "./models/lumpsumService.mocks.js";
export * from "./models/nonbusinessDay.mocks.js";
export * from "./models/project.mocks.js";
// export * from "./models/service.mocks.js";
export * from "./models/targethours.mocks.js";
export * from "./models/team.mocks.js";
export * from "./models/user.mocks.js";
// export * from "./models/userReport.mocks.js";
export * from "./models/worktimeRegulation.mocks.js";
export * from "./models/workTimes.mocks.js";

/**
 * Sets the seed for Faker.js (which is used by the mocks).
 *
 * @see https://github.com/Marak/Faker.js
 */
export const setFakerSeed = (seed: number) => {
  faker.seed(seed);
};
