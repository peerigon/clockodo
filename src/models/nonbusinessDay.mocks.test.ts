import { describe, expect, test } from "vitest";
import { createNonbusinessDayMocks } from "./nonbusinessDay.mocks.js";

describe("createNonbusinessDayMocks()", () => {
  test("It creates realistic mocks", () => {
    expect(createNonbusinessDayMocks({ count: 20 })).toMatchSnapshot();
  });

  test("It re-uses ids for the next year", () => {
    const [nonbusinessDay1, nonbusinessDay2] = createNonbusinessDayMocks({
      count: 2,
      dateBetween: [new Date(2020, 11, 31), new Date(2021, 0, 2)],
    });

    expect(nonbusinessDay1!.id === nonbusinessDay2!.id);
  });
});
