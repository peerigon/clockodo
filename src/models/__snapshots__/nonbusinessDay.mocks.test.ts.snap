// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNonbusinessDayMocks() It creates realistic mocks 1`] = `
[
  {
    "date": "2020-02-21",
    "halfDay": false,
    "id": 0,
    "name": "similique libero ad",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-03-24",
    "halfDay": false,
    "id": 1,
    "name": "velit ipsa quidem",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-04-14",
    "halfDay": false,
    "id": 2,
    "name": "perspiciatis cum similique",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-05-05",
    "halfDay": false,
    "id": 3,
    "name": "reprehenderit dolores veritatis",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-05-23",
    "halfDay": false,
    "id": 4,
    "name": "expedita voluptatum vel",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-05-26",
    "halfDay": false,
    "id": 5,
    "name": "magnam illum dignissimos",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-05-30",
    "halfDay": false,
    "id": 6,
    "name": "cum iste corporis",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-06-03",
    "halfDay": false,
    "id": 7,
    "name": "commodi ad animi",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-06-05",
    "halfDay": false,
    "id": 8,
    "name": "voluptas amet consectetur",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-06-25",
    "halfDay": false,
    "id": 9,
    "name": "doloribus similique cumque",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-06-28",
    "halfDay": false,
    "id": 10,
    "name": "quo eum voluptatem",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-07-20",
    "halfDay": false,
    "id": 11,
    "name": "dolores dignissimos quia",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-07-31",
    "halfDay": false,
    "id": 12,
    "name": "inventore quaerat officiis",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-09-07",
    "halfDay": true,
    "id": 13,
    "name": "nobis debitis quidem",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-09-09",
    "halfDay": false,
    "id": 14,
    "name": "laudantium blanditiis quam",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-09-11",
    "halfDay": false,
    "id": 15,
    "name": "excepturi dicta aliquam",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-09-17",
    "halfDay": true,
    "id": 16,
    "name": "ducimus commodi fuga",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-09-20",
    "halfDay": false,
    "id": 17,
    "name": "voluptates non beatae",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-10-12",
    "halfDay": false,
    "id": 18,
    "name": "ducimus doloribus consectetur",
    "nonbusinessgroupsId": 0,
  },
  {
    "date": "2020-12-24",
    "halfDay": false,
    "id": 19,
    "name": "iure iste nobis",
    "nonbusinessgroupsId": 0,
  },
]
`;
